### ISSUE TITLE
VDS.RDF.Parsing.RdfParseException with unexpected rdf:parseType attribute

### ISSUE THREAD

**Issue by glebkozh** [dotnetrdf/dotnetrdf#202]
Hello!
I've encountered the following issue.
This rdf parsing code fails with an exception:

```
var content = File.ReadAllText(@"example.owl");
var graph = new VDS.RDF.Graph();
var xml = new XmlDocument();
xml.LoadXml(content);
var parser = new RdfXmlParser();
parser.Load(graph, xml);
```
The exception is following: 

```
An unhandled exception of type 'VDS.RDF.Parsing.RdfParseException' occurred in dotNetRDF.dll

Additional information: Unexpected Attribute 'rdf:parseType' was encountered on a Property Element with Parse Type 'Collection'.  Only rdf:ID and rdf:parseType are allowed on Property Elements with Parse Type 'Collection'

Occurred in Grammar Production 'Parse Type Collection Property Element'

[Source XML]

<owl:oneOf rdf:parseType="Collection" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:owl="http://www.w3.org/2002/07/owl#"><owl:Thing rdf:about="http://www.somecompany.com/ns/A#C" /><owl:Thing rdf:about="http://www.somecompany.com/ns/A#D" /></owl:oneOf>
```

StackTrace:

```
   at VDS.RDF.Parsing.RdfXmlParser.GrammarProductionParseTypeCollectionPropertyElement(RdfXmlParserContext context, IEventQueue`1 eventlist, IRdfXmlEvent parent)
   at VDS.RDF.Parsing.RdfXmlParser.GrammarProductionPropertyElement(RdfXmlParserContext context, IEventQueue`1 eventlist, IRdfXmlEvent parent)
   at VDS.RDF.Parsing.RdfXmlParser.GrammarProductionPropertyElementList(RdfXmlParserContext context, IEventQueue`1 eventlist, IRdfXmlEvent parent)
   at VDS.RDF.Parsing.RdfXmlParser.GrammarProductionNodeElement(RdfXmlParserContext context, IEventQueue`1 eventlist)
   at VDS.RDF.Parsing.RdfXmlParser.GrammarProductionNodeElementList(RdfXmlParserContext context, IEventQueue`1 eventlist)
   at VDS.RDF.Parsing.RdfXmlParser.GrammarProductionRDF(RdfXmlParserContext context, ElementEvent element)
   at VDS.RDF.Parsing.RdfXmlParser.GrammarProductionDoc(RdfXmlParserContext context, RootEvent root)
   at VDS.RDF.Parsing.RdfXmlParser.ProcessEventQueue(RdfXmlParserContext context)
   at VDS.RDF.Parsing.RdfXmlParser.Parse(RdfXmlParserContext context)
   at VDS.RDF.Parsing.RdfXmlParser.Load(IGraph g, XmlDocument document)
```
Issue arises in `VDS.RDF.Parsing.RdfXmlParser.GrammarProductionParseTypeCollectionPropertyElement` method, because condition `else if (a.Namespace.Equals(NamespaceMapper.RDF) && a.LocalName.Equals("parseType"))` (line 1373) happens to evaluate to `false`. `NamespaceMapper.RDF` has a constant value of full namespace `"http://www.w3.org/1999/02/22-rdf-syntax-ns#"` , while attribute's `a.Namespace` contains local value `"rdf"`. They do not match and lead to exception described above.

This happens in dotnetrdf version 2.1.0 and vds.common 1.10.0. Previously I used dotnetrdf version 1.0.10 and vds.common version 1.6.4. In older version there are no separate checks for namespace and localname, but there is single check for `"rdf:parseType"`, which had been pretty valid.
example.owl content:

```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:owl="http://www.w3.org/2002/07/owl#">
	<owl:Class rdf:about="http://www.somecompany.com/ns/A#B">
		<owl:oneOf rdf:parseType="Collection">
			<owl:Thing rdf:about="http://www.somecompany.com/ns/A#C"/>
			<owl:Thing rdf:about="http://www.somecompany.com/ns/A#D"/>
		</owl:oneOf>
	</owl:Class>
</rdf:RDF>
```
Thanks!

**Comment by langsamu** [comment]
### TL;DR
For a workaround use streaming parsing instead of DOM:
```c#
var content = File.ReadAllText(@"example.owl");
var graph = new VDS.RDF.Graph();
var parser = new RdfXmlParser();
parser.Load(graph, new StringReader(content));
```
or
```c#
var graph = new VDS.RDF.Graph();
graph.LoadFromFile("example.owl");
```
### Background
The change in `GrammarProductionParseTypeCollectionPropertyElement` happened in #163. That PR changed how the RDF/XML parser handles XML element and attribute names. While the previous version relied on QNames with the prefix `rdf` (i.e. `rdf:parseType`), the new one relies on the actual namespace URI of the node (i.e. `{http://www.w3.org/1999/02/22-rdf-syntax-ns#}parseType`.

This is a good thing as XML documents can have whatever prefix defined for the RDF namespace URI and still be correct. The [RDF/XML test suite](https://www.w3.org/2013/RDFXMLTests/) actually has a test (rdf-ns-prefix-confusion-test0014) for this case, which the previous version failed on (#140).

### Issue
Unfortunately the fix didn't account for DOM parsing mode.
In [`RdfXmlParser`](https://github.com/dotnetrdf/dotnetrdf/blob/master/Libraries/dotNetRDF/Parsing/RDFXMLParser.cs#L86)
When [loading from an `XmlDocument`](https://github.com/dotnetrdf/dotnetrdf/blob/master/Libraries/dotNetRDF/Parsing/RDFXMLParser.cs#L257) or [constructing with `RdfXmlParserMode.DOM`](https://github.com/dotnetrdf/dotnetrdf/blob/master/Libraries/dotNetRDF/Parsing/RDFXMLParser.cs#L124),  the [parser context](https://github.com/dotnetrdf/dotnetrdf/blob/master/Libraries/dotNetRDF/Parsing/Contexts/RdfXmlParserContext.cs#L40) uses the `DomBasedEventGenerator` which [fires events with `Namespace` set to the prefix of the attribute, not its namespace URI](https://github.com/dotnetrdf/dotnetrdf/blob/5b7fc480346c90eb9b164fb4f8ee09f378442d52/Libraries/dotNetRDF/Parsing/Events/RdfXml/DomBasedEventGenerator.cs#L312).
This fails the changes in #163.

### Solution
I had a look at modifying [`AttributeEvent `](https://github.com/dotnetrdf/dotnetrdf/blob/master/Libraries/dotNetRDF/Parsing/Events/RdfXml/XMLEvents.cs#L447) and basing all of it on namespace URI and local name (instead of prefix and local name).
This would require more changes to the parser, I think that's fine.
However [`RdfXmlSpecHelper`](https://github.com/dotnetrdf/dotnetrdf/blob/master/Libraries/dotNetRDF/Parsing/RdfXmlSpecsHelper.cs) relies heavily on QNames, I'm not sure what should be done about that. @kal @tpluscode?

**Comment by kal** [comment]
Thanks for the report @glebkozh and thanks for the analysis @langsamu!

It seems to me that the right solution is the one you suggest (as is the workaround).

AttributeEvent should already be surfacing the Namespace and LocalName properties as well as QName, so I think that the fixes are mostly around RdfXmlParser and RdfXmlSpecHelper.

To avoid changing the public API of RdfXmlSpecHelper, we could add overloads of most methods that currently take a qname (the overload would take a namespace URI parameter and a local name parameter). The methods that currently take an AttributeEvent I think could just be reimplemented to test the Namespace and LocalName properties rather than testing the QName property.

It would be a non-trivial bit of work to do and ideally I would like to get a batch of parser tests that exhibit this issue into the test suite before starting work on it. That said, I think it is probably something that I could get to grips with over the next few days.
